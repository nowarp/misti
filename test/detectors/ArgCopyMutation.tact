const WHATEVER: Int = 0;

contract Test {
    m: map<Int, Int>;
    fun test(a: Int, m: map<Int, Int>) {
        m.set(3, 3); // Bad

        // OK: No warning since nobody will expect an integral type to be modified
        a += 1;
        a = 42;

        // OK: Accessing map element
        let _b: Int? = m.get(3);

        // OK: Modifying contract fields
        self.m.set(3, 3);
    }

    // To ensure we report the argument once per function
    fun testOnce(m: map<Int, Int>) {
        m.set(1, 1);
        m.set(2, 2);
    }

    // Test the warning position: it should point out to `set`
    fun testForeach(m: map<Int, Int>) {
        foreach(key, value in m) { // No warning here
            m.set(3, 3); // Bad
        }
    }

    fun id(m: map<Int, Int>): map<Int, Int> {
        return m;
    }

    // OK: The modified argument returned from all the branches
    fun testReturn1(m: map<Int, Int>): map<Int, Int> {
        m.set(3, 3); // OK
        if (WHATEVER == 0) {
            return m;
        } else {
            return self.id(m);
        }
    }
    // Bad: One of the branches returns something else
    fun testReturn2(m: map<Int, Int>): map<Int, Int> {
        m.set(3, 3); // Bad
        if (WHATEVER == 0) {
            return m;
        } else {
            let mCopy: map<Int, Int> = m;
            return mCopy; // Bad
        }
    }
}
