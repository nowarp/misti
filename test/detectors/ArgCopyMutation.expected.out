[HIGH] ArgCopyMutation: Function test argument m is mutated
test/detectors/ArgCopyMutation.tact:6:9:
  5 |     fun test(a: Int, m: map<Int, Int>) {
> 6 |         m.set(3, 3); // Bad
              ^
  7 | 
Mutating function arguments has no effect outside the function due to call-by-value semantics
Help: Return the modified value or use the contract's state to avoid unnecessary mutations
See: https://nowarp.io/tools/misti/docs/detectors/ArgCopyMutation

[HIGH] ArgCopyMutation: Function testOnce argument m is mutated (1 more times)
test/detectors/ArgCopyMutation.tact:21:9:
  20 |     fun testOnce(m: map<Int, Int>) {
> 21 |         m.set(1, 1);
               ^
  22 |         m.set(2, 2);
Mutating function arguments has no effect outside the function due to call-by-value semantics
Help: Return the modified value or use the contract's state to avoid unnecessary mutations
See: https://nowarp.io/tools/misti/docs/detectors/ArgCopyMutation

[HIGH] ArgCopyMutation: Function testForeach argument m is mutated
test/detectors/ArgCopyMutation.tact:28:13:
  27 |         foreach(key, value in m) { // No warning here
> 28 |             m.set(3, 3); // Bad
                   ^
  29 |         }
Mutating function arguments has no effect outside the function due to call-by-value semantics
Help: Return the modified value or use the contract's state to avoid unnecessary mutations
See: https://nowarp.io/tools/misti/docs/detectors/ArgCopyMutation

[HIGH] ArgCopyMutation: Function testReturn2 argument m is mutated
test/detectors/ArgCopyMutation.tact:47:9:
  46 |     fun testReturn2(m: map<Int, Int>): map<Int, Int> {
> 47 |         m.set(3, 3); // Bad
               ^
  48 |         if (WHATEVER == 0) {
Mutating function arguments has no effect outside the function due to call-by-value semantics
Help: Return the modified value or use the contract's state to avoid unnecessary mutations
See: https://nowarp.io/tools/misti/docs/detectors/ArgCopyMutation